### ISSUE TITLE
Infinite redirection bug due to language code parsing issue

### ISSUE THREAD

**Issue by matthieu-gll** [QuiiBz/next-international#96]
## Issue Description:
I am encountering an infinite redirection bug in my Next.js application, which is caused by a language code parsing issue. The application uses the external library, "next-international," to handle internationalization.

## Reproduction Steps:

Set up a Next.js application with internationalization using the "next-international" library.
Use language codes containing a hyphen ("-"), e.g., "en-US" and "fr-FR," to represent different locales.

### Expected Behavior:
The application should successfully identify and use the correct locale based on the "Accept-Language" header or the "Next-Locale" cookie.

### Actual Behavior:
The "negotiateAcceptLanguage" function does not correctly parse the "Accept-Language" header when the language codes contain a hyphen ("-"). As a result, it retrieves only the first part of the language code and ignores the region-specific part (e.g., "fr" instead of "fr-FR").

## Issue Analysis:
The root cause of the bug lies in the "negotiateAcceptLanguage" function's splitting logic for the "Accept-Language" header. When the function attempts to split the header by a comma (","), it only retrieves the first part before the hyphen ("-"). This leads to incorrect locale identification and may cause an infinite redirection loop when handling internationalized routes.

## Code Snippet with the Issue:

```
// File: src/app/middleware/index.ts

function negotiateAcceptLanguage(request: NextRequest) {
    const header = request.headers.get('Accept-Language');
    console.log(header);
    const locale = header?.split(',')?.[0]; // Incorrect splitting logic for language codes with hyphen ("-")
    return locale ?? null;
}

// ... Rest of the code ...

```
## Proposed Solution:
To resolve the infinite redirection issue, the library's parsing logic for language codes should be modified to handle hyphens ("-") correctly. This can be achieved by using a different approach to split the "Accept-Language" header, allowing it to retrieve the complete language code, including the region-specific part.

Additional Information:

Next.js version: 13.4.12
next-international version: ^0.8.1

If it's good for you, I'll be happy to help you by doing the PR, indeed, this would be one of my first contributions to an open source project, especially since I really appreciate your project, because it taught me a lot about how the core of next.js works and it saved me a lot of time.

In all cases, I hope this information helps in resolving the issue. Please let me know if you need any further details or assistance with the bug fix. Thank you!

Best regards,

**Comment by QuiiBz** [comment]
Thanks for this detailed explanation. It's indeed intended to parse the language code by `-` to only retrieve the locale, without the region-specific part. Now I'm wondering if there is a need to differentiate between regions within the same locale (e.g `fr-FR` and `fr-BE`, `en-US` and `en-CA`)? What's the reason you can't use `fr` / `en` directly?

**Comment by SamAmann** [comment]
I believe the reason would be to allow adding region specificity to a language. So, it's not that you "can't" use `fr `/ `en` directly, it's you "could" use more specific ones, using the region convention.
Anyway, I'm still wondering how that would render in the `/locales/[lang].json` file: 
would we have to create "enGB.json" and "enUS.json" files?
or could we add specific translation in each language file?

I'll +1 on the incredible work you have pulled off! Amazing quality and time saver!
Would be happy to help out as well!

**Comment by Yovach** [comment]
> Thanks for this detailed explanation. It's indeed intended to parse the language code by `-` to only retrieve the locale, without the region-specific part. Now I'm wondering if there is a need to differentiate between regions within the same locale (e.g `fr-FR` and `fr-BE`, `en-US` and `en-CA`)? What's the reason you can't use `fr` / `en` directly?
> 
> 

I think this is an interesting feature, as some regions use different spellings for words like "color" and "colour" in US and UK.

**Comment by QuiiBz** [comment]
When we introduced the `rewrite` strategy in the middleware, we wanted to add a new option to override the way that a locale is determined from a `Request`. But this feature didn't land because it wasn't really needed at that time: https://github.com/QuiiBz/next-international/pull/83#discussion_r1278893308

Could your issue be solved if we add this `contentNegotiator` option to the i18n middleware, which will be called when looking for the locale of a `Request`? If yes, feel free to send a PR!

**Comment by matthieu-gll** [comment]
> Thanks for this detailed explanation. It's indeed intended to parse the language code by `-` to only retrieve the locale, without the region-specific part. Now I'm wondering if there is a need to differentiate between regions within the same locale (e.g `fr-FR` and `fr-BE`, `en-US` and `en-CA`)? What's the reason you can't use `fr` / `en` directly?

This is a specific need for my application, to put it simply, my web application brings great added value to the content created by its community thanks in particular to advanced translations. For the moment being in the mvp phase I can be satisfied with languages like fr / en but in the future there is a good chance that it will be a required feature of my application

**Comment by QuiiBz** [comment]
I've released [0.9.0](https://github.com/QuiiBz/next-international/releases/tag/0.9.0) which includes a new `resolveLocaleFromRequest` option inside `createI18nMiddleware`, where you can add your own locale resolution logic. 

Feel free to open another issue if you're not able to make it work.
